wit/types.wit

```diff
@@ -217,72 +223,42 @@ interface types {
   /// Trailers is an alias for Fields.
   type trailers = fields;
 
-  /// Represents an incoming HTTP Request.
-  resource incoming-request {
-
-    /// Returns the method of the incoming request.
-    method: func() -> method;
-
-    /// Returns the path with query parameters from the request, as a string.
-    path-with-query: func() -> option<string>;
-
-    /// Returns the protocol scheme from the request.
-    scheme: func() -> option<scheme>;
-
-    /// Returns the authority from the request, if it was present.
-    authority: func() -> option<string>;
-
-    /// Get the `headers` associated with the request.
-    ///
-    /// The returned `headers` resource is immutable: `set`, `append`, and
-    /// `delete` operations will fail with `header-error.immutable`.
-    ///
-    /// The `headers` returned are a child resource: it must be dropped before
-    /// the parent `incoming-request` is dropped. Dropping this
-    /// `incoming-request` before all children are dropped will trap.
-    headers: func() -> headers;
-
-    /// Gives the `incoming-body` associated with this request. Will only
-    /// return success at most once, and subsequent calls will return error.
-    consume: func() -> result<incoming-body>;
-  }
-
-  /// Represents an outgoing HTTP Request.
-  resource outgoing-request {
+  /// Represents an HTTP Request.
+  resource request {
 
-    /// Construct a new `outgoing-request` with a default `method` of `GET`, and
+    /// Construct a new `request` with a default `method` of `GET`, and
     /// `none` values for `path-with-query`, `scheme`, and `authority`.
     ///
-    /// * `headers` is the HTTP Headers for the Request.
+    /// * `body` is the contents of the body, as a stream of bytes.
+    /// * `headers` is the HTTP Headers for the Response.
+    /// * `trailers` is a `future` which resolves to the HTTP Trailers for the
+    ///   Response.
+    /// * `options` is optional `request-options` to be used if the request is
+    ///   sent over a network connection.
     ///
     /// It is possible to construct, or manipulate with the accessor functions
-    /// below, an `outgoing-request` with an invalid combination of `scheme`
+    /// below, an `request` with an invalid combination of `scheme`
     /// and `authority`, or `headers` which are not permitted to be sent.
-    /// It is the obligation of the `outgoing-handler.handle` implementation
-    /// to reject invalid constructions of `outgoing-request`.
+    /// It is the obligation of the `handler.handle` implementation
+    /// to reject invalid constructions of `request`.
     constructor(
-      headers: headers
+      body: stream<u8>,
+      headers: headers,
+      trailers: future<trailers>,
```

Comment:
```suggestion
      trailers: option<future<trailers>>,
```

Reply:
Yeah, I originally wrapped it in an `option`, then removed it, thinking it should be more symmetric with the `headers` parameter; i.e. if you have no headers, just pass an empty `headers`, and if you have no trailers (or don't know yet whether you'll have any), you pass a `future` that (may) resolve to an empty `trailers`.  In other words, if we make `trailers` optional, should we also make `headers` optional?

Reply:
Fair point and I wondered the same.  Here, I was mostly just matching 0.2 (which made trailers optional, but headers non-optional).  But perhaps you could make the argument that you almost always want headers but trailers are pretty rare, but maybe for symmetry both should be optional.

Reply:
Yeah, trailers _are_ rare, and forcing everyone to create and pass a future would be annoying.  I'll go ahead and make them optional and leave the headers required for now.

Reply:
We might even want to remove the `trailers` param from the constructor and add a setter instead.  How does that sound?

Reply:
There's an interesting consequence to that: passing a `future` to a method (as opposed to ctor) holds onto the borrowed self-request handle passed to the method-call until the future is resolved which in turn prevents you from transferring ownership of the request out of the current component (which one naturally wants to do before the trailers are resolved).  The ctor doesn't have this problem b/c the ctor doesn't borrow anything.

(As a side note: in most cases, this borrow-lasts-the-whole-async-call rule makes sense (e.g., I shouldn't be transferring ownership of a file in the middle of doing a streaming `write` method call), but in this case it maybe seems restrictive.  If this is a problem, there are various ways to avoid this limitation (in the design or using an intermediate constructed resource type) that we can discuss.)

Reply:
Yeah, I was wondering about the borrow.  Rust `async` functions have the same property, which you can work around by writing a non-async function that returns a `Future` with a `'static` lifetime, but that trick won't work in our "colorless" approach (which is fine).

Anyway, let's just leave it as a constructor param for now.

Reply:
Interesting!  I suppose the "intermediate constructed resource type" workaround is somewhat analogous to the `'static` workaround.  That is, if we really didn't want a constructor parameter, we could have:
```wit
resource future-trailers {
  constructor(f: future<trailers>);
}
resource request {
  set-trailers: func(f: future-trailers);
}
```
If this became a necessary or common workaround, I wonder if we couldn't come up with some more first-class support for this technique that avoided the superfluous resource type.  But maybe the constructor-parameter approach ends up being good enough for most cases.  I guess we'll have to see what happens.

Reply:
Would something like this work?
```
resource request {
  constructor(headers: headers, body: stream<u8>);
  set-trailers: static func(request: request, trailers: future<trailers>) -> request;
  ...
}
```